<a href="https://colab.research.google.com/github/pawan-cpu/Learn-Python-with-Pawan-Kumar/blob/main/Copy_of_2021_12_24_pawan_lesson34.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 34: Time Series Data


|Particulars|Description|
|-|-|
|**Topics Covered**|Missing Values|
||Parsing `datetime` Values|
|||
|**Lesson Description**|A student will learn to extract day, month, year, hours, minutes and seconds in a time-series data.|
|||
|**Lesson Duration**|45 minutes|
|||
|**Learning Outcomes**|Convert a datetime value, reported as an object in a Pandas series, to a `datetime` object|
||Format a date in different commonly used date formats such as DD-MM-YYYY|
||Extract day, month, year, hours, minutes and seconds from a `datetime` value in a time-series data|
|||


---

### Teacher-Student Tasks

From this class onwards, we will work on a new project in which we have to analyse data on air quality in one of the cities in Italy. We need to determine whether the concentration of air pollutants increase or decrease over time. 

In the process, we will also learn how to work with time-series data, i.e, how to work with date and time values. Towards the end of the project, we need to create time-series plots such as line plots, bar plots and multivariate boxplots to observe the trend of air pollution concentration over time.

Now, skim through the data description to understand the kind of features and corresponding values we have in this data set on air quality.

---

### Data Description

The dataset contains 9358 instances of hourly averaged responses from an array of 5 metal oxide chemical sensors embedded in an Air Quality Chemical Multisensor Device.

The device was located on the field in a significantly polluted area, at road level, within an Italian city. Data were recorded from March 2004 to April 2005 (one year) representing the longest freely available recordings of on field-deployed air quality chemical sensor devices responses. A co-located reference certified analyzer provided **Ground Truth (GT)** hourly averaged concentrations for Carbon Monoxide ($\text{CO}$), Non-Methane Hydrocarbons ($\text{NMHC}$), Benzene ($\text{C}_6\text{H}_6$), total Nitrogen Oxides ($\text{NO}_x$) & Nitrogen Dioxide ($\text{NO}_2$), and Ozone ($\text{O}_3$). These chemical compounds are commonly occurring air pollutants.

|Index|Columns|Description|
|-|-|-|
|0|Date|Date (DD/MM/YYYY)|
|1|Time|Time (HH.MM.SS)|
|2|CO(GT)|True hourly averaged concentration $\text{CO}$ in $\frac{mg}{m^3}$ (reference analyzer)|
|3|PT08.S1(CO)|PT08.S1 (tin oxide) hourly averaged sensor response (nominally $\text{CO}$ targeted)|
|4|NMHC(GT)|True hourly averaged overall Non-Methane Hydrocarbons concentration in $\frac{\mu g}{m^3}$|
|5|C6H6(GT)|True hourly averaged Benzene concentration in $\frac{\mu g}{m^3}$||
|6|PT08.S2(NMHC)|PT08.S2 (titania) hourly averaged sensor response (nominally $\text{NMHC}$ targeted)|
|7|NOx(GT)|True hourly averaged $\text{NO}_x$ concentration in Parts per billion (ppb)|
|8|PT08.S3(NOx)|PT08.S3 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_x$ targeted)|
|9|NO2(GT) |True hourly averaged $\text{NO}_2$ concentration in $\frac{\mu g}{m^3}$|
|10|PT08.S4(NO2)|PT08.S4 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_2$ targeted)|
|11|PT08.S5(O3) |PT08.S5 (indium oxide) hourly averaged sensor response (nominally $\text{O}_3$ targeted)|
|12|T|Temperature in °C|
|13|RH|Relative Humidity (%)|
|14|AH|AH Absolute Humidity|

The above dataset contains missing (or null) values and `-200` as garbage values.

**Dataset Credits:** https://archive.ics.uci.edu/ml/datasets/Air+quality

**Citation**: S. De Vito, E. Massera, M. Piga, L. Martinotto, G. Di Francia.

On field calibration of an electronic nose for benzene estimation in an urban pollution monitoring scenario, Sensors and Actuators B: Chemical, Volume 129.

---

### About Air Pollutants

Please note that you don't need to go through the entire description to solve this project. It is provided to give a brief description on common air pollutants. You may choose to remove this text cell before teaching a student.

**Carbon Monoxide** ($\text{CO}$)

It is a poisonous gas. It reacts with haemoglobin because of which it is unable to supply oxygen to the human body. At very high concentrations of carbon monoxide, up to 40% of the haemoglobin can bind with carbon monoxide which is enough to certainly kill humans. Carbon monoxide is also known as the silent killer because it has no colour, taste or smell.


**Benzene ($\text{C}_6\text{H}_6$)**

It is a colorless or light yellow liquid at room temperature. It has a sweet odour and is highly flammable. It evaporates into the air very quickly, but its vapour is heavier than air and may sink into low-lying areas. Outdoor air contains low levels of benzene from tobacco smoke, gas stations, motor vehicle exhaust, and industrial emissions. Indoor air generally contains levels of benzene higher than those in outdoor air. The benzene in indoor air comes from products that contain benzene such as glues, paints, furniture wax, and detergents. The air around hazardous waste sites or gas stations can contain higher levels of benzene than in other areas. Tobacco smoke is also a major source of benzene.

People who breathe in high levels of benzene may develop drowsiness, dizziness, rapid or irregular heartbeat, headaches, tremors, confusion,  unconsciousness and death (at very high levels)

**Nitrogen Oxides ($\text{NO}_x$) & Nitrogen Dioxide ($\text{NO}_2$)** 

Nitrogen dioxide and nitric oxide together are referred to as oxides of nitrogen ($\text{NO}_x$).

Nitrogen dioxide is an irritant gas. When nitrogen is released during fuel combustion, it combines with oxygen atoms to create nitric oxide ($\text{NO}$). This further combines with oxygen to create nitrogen dioxide ($\text{NO}_2$). Nitric oxide is not considered to be hazardous to health at typical ambient concentrations, but nitrogen dioxide can be. 

$\text{NO}_x$ gases react to form smog and acid rain as well as being central to the formation of fine particles (PM) and ground level ozone, both of which are associated with adverse health effects.

**Ozone ($\text{O}_3$)**

Ozone is a colourless and odourless gas. Ground-level ozone forms when pollutants from cars and trucks, power plants, factories, and other sources come in contact with each other in heat and sunlight. Ground-level ozone is one of the biggest parts of smog, and it is usually worse in the summer months.

Many scientific studies have linked ground-level ozone contact to such varied problems as

- asthma, bronchitis, and emphysema

- pneumonia or bronchitis

- lung and throat irritation


**Non-Methane Hydrocarbons ($\text{NMHC}$)**

All the hydrocarbons except for methane ($\text{CH}_4$) are collectively called Non-Methane Hydrocarbons. The elevated ambient concentration of NMHCs in an urban environment has a significant impact on climate change and human health.

Hydrocarbons (or organic compounds) are naturally occurring substances that can be found in all living things. The solvents being emitted from manufacturing processes are considered Volatile Organic Compounds (VOC). The hydrocarbons are considered volatile since they produce a vapour at room temperature and normal atmospheric pressure. VOCs are contributors to the formation of ozone and smog.  

The mechanism for ozone production is:

$$\text{VOC}  +  \text{NO}_x  +  \text{sunlight (UV)} \longrightarrow  \text{O}_3$$

Hydrocarbons are the main component of crude oil, natural gases, and most pesticides. All of these substances contribute to the greenhouse effect, and the depletion of the ozone layer. They also reduce the photosynthetic ability of plants, increase cancer rates in humans and animals, and increase the risk of respiratory illness. 


---

#### Task 1: Loading Data

Let's begin the class with importing the required module and importing the time-series dataset on air quality. Here's the dataset link:

https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/air-quality/AirQualityUCI.csv

The values in the dataset are separated by semi-columns (`;`) instead of commas. The image below shows the first 10 lines of the CSV file.

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/semi_colon_sv.png'>

To read the given CSV file, pass the `sep = ';'` parameter inside the `read_csv()` function along with the file location.

In [ ]:
# S1.1: Import the required modules and load the time-series dataset on air quality. Also, display the first five rows.
import pandas as pd
df=pd.read_csv('https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/air-quality/AirQualityUCI.csv',sep=';')
df.head()

Date      Time CO(GT)  ...      AH  Unnamed: 15 Unnamed: 16
0  10/03/2004  18.00.00    2,6  ...  0,7578          NaN         NaN
1  10/03/2004  19.00.00      2  ...  0,7255          NaN         NaN
2  10/03/2004  20.00.00    2,2  ...  0,7502          NaN         NaN
3  10/03/2004  21.00.00    2,2  ...  0,7867          NaN         NaN
4  10/03/2004  22.00.00    1,6  ...  0,7888          NaN         NaN

[5 rows x 17 columns]

The dataset contains two unnamed columns. They are `Unnamed: 15` and `Unnamed: 16`. Both the columns contain `NaN` or null values. We need to drop them. Also, few of the columns contain numeric values separated by comma. We need to replace the commas with periods (or dots).

Let's first look at the complete information on the `df` DataFrame.

In [ ]:
# S1.2: Apply the 'info()' function on the 'df' DataFrame.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   object 
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   object 
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   object 
 13  RH             9357 non-null   object 
 14  AH             9357 non-null   object 
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(10), object(7)
memory usage: 1.2+ MB


The values in the `Date` and `Time` columns need to be converted to the `datetime` values. Additionally, the values in the `CO(GT), C6H6(GT), T, RH` and `AH` columns need to be converted to float-point numbers.

---

#### Task 2: Missing Values

We already know that the `Unnamed: 15 and Unnamed: 16` columns contain the missing (or null) values. Let's find out if more columns also contain the null values.

In [ ]:
# S2.1: Check for the missing values in the 'df' DataFrame.
df.isnull().sum()

Date              114
Time              114
CO(GT)            114
PT08.S1(CO)       114
NMHC(GT)          114
C6H6(GT)          114
PT08.S2(NMHC)     114
NOx(GT)           114
PT08.S3(NOx)      114
NO2(GT)           114
PT08.S4(NO2)      114
PT08.S5(O3)       114
T                 114
RH                114
AH                114
Unnamed: 15      9471
Unnamed: 16      9471
dtype: int64

All the values contained in the `Unnamed: 15 and Unnamed: 16` columns are the missing (or null) values. The other columns contain 114 null values. So, let's drop the last two columns using the `drop()` function.

It removes rows or columns by specifying label names and corresponding axis, or by specifying directly index or column names. In this case, we will specify the column names.

In [ ]:
# S2.2: Drop the 'Unnamed: 15 and Unnamed: 16' columns from the 'df' DataFrame.
df = df.drop(columns = ["Unnamed: 15","Unnamed: 16"] ,axis = 1 )
df.head()

Date      Time CO(GT)  PT08.S1(CO)  ...  PT08.S5(O3)     T    RH      AH
0  10/03/2004  18.00.00    2,6       1360.0  ...       1268.0  13,6  48,9  0,7578
1  10/03/2004  19.00.00      2       1292.0  ...        972.0  13,3  47,7  0,7255
2  10/03/2004  20.00.00    2,2       1402.0  ...       1074.0  11,9  54,0  0,7502
3  10/03/2004  21.00.00    2,2       1376.0  ...       1203.0  11,0  60,0  0,7867
4  10/03/2004  22.00.00    1,6       1272.0  ...       1110.0  11,2  59,6  0,7888

[5 rows x 15 columns]

The `axis=1` parameter tells Python to drop columns. In this case, if you don't pass the `axis=1` parameter, Python will drop the given columns because we have specified the column names to be dropped. Nonetheless, let's follow a good habit of specifying whether we want to remove a column or a row.

Now, let's verify whether the `Unnamed: 15 and Unnamed: 16` columns are removed or not from the `df` DataFrame by printing all the column names.

In [ ]:
# S2.3: Get the list of columns present in the 'df' DataFrame after removing the 'Unnamed: 15 and Unnamed: 16' columns.
df.columns

Index(['Date', 'Time', 'CO(GT)', 'PT08.S1(CO)', 'NMHC(GT)', 'C6H6(GT)',
       'PT08.S2(NMHC)', 'NOx(GT)', 'PT08.S3(NOx)', 'NO2(GT)', 'PT08.S4(NO2)',
       'PT08.S5(O3)', 'T', 'RH', 'AH'],
      dtype='object')

We know that the other columns have 114 missing values. Let's see which rows in the `Date` column are missing.

In [ ]:
# S2.4: Print the rows missing in the 'Date' column.
df[df['Date'].isnull()==True]


Date Time CO(GT)  PT08.S1(CO)  ...  PT08.S5(O3)    T   RH   AH
9357  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9358  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9359  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9360  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9361  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
...   ...  ...    ...          ...  ...          ...  ...  ...  ...
9466  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9467  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9468  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9469  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN
9470  NaN  NaN    NaN          NaN  ...          NaN  NaN  NaN  NaN

[114 rows x 15 columns]

We can see that all the rows starting from index `9357` to `9470` contain missing values in all the columns. 

**The `dropna()` Function**

Let's drop them all using the `dropna()` function. It removes the rows or columns containing the missing values. If no parameter is passed to the `dropna()` function, then it removes all the rows containing at least one null value.

In [ ]:
# S2.5: Drop the rows containing at least one null value in the 'df' DataFrame using the 'dropna()' function.
df=df.dropna()


Now, let's check whether all the missing values are removed or not.

In [ ]:
# S2.6: Check whether all the missing values are removed or not.
df.isnull().sum()

Date             0
Time             0
CO(GT)           0
PT08.S1(CO)      0
NMHC(GT)         0
C6H6(GT)         0
PT08.S2(NMHC)    0
NOx(GT)          0
PT08.S3(NOx)     0
NO2(GT)          0
PT08.S4(NO2)     0
PT08.S5(O3)      0
T                0
RH               0
AH               0
dtype: int64

Finally, let's find out how many rows and columns are in the DataFrame after removing the rows and columns containing the missing values. 

In [ ]:
# S2.7: How many rows and columns are in the DataFrame after removing the rows and columns containing the missing values?
df.shape

(9357, 15)

Let's also see the last 5 rows of the `df` DataFrame. Their indices should be `9352` to `9356`.

In [ ]:
# S2.8: Print the last 5 rows of the 'df' DataFrame.
df.tail()

Date      Time CO(GT)  PT08.S1(CO)  ...  PT08.S5(O3)     T    RH      AH
9352  04/04/2005  10.00.00    3,1       1314.0  ...       1729.0  21,9  29,3  0,7568
9353  04/04/2005  11.00.00    2,4       1163.0  ...       1269.0  24,3  23,7  0,7119
9354  04/04/2005  12.00.00    2,4       1142.0  ...       1092.0  26,9  18,3  0,6406
9355  04/04/2005  13.00.00    2,1       1003.0  ...        770.0  28,3  13,5  0,5139
9356  04/04/2005  14.00.00    2,2       1071.0  ...        816.0  28,5  13,1  0,5028

[5 rows x 15 columns]

---

#### Task 3: Parsing `datetime` Values - `pd.to_datetime()` Function

Now, we need to convert the values stored in the `Date` and `Time` columns to `datetime` values so that later we can sort the DataFrame in the chronological order. A `datetime` value is a combination of date and time.

Let's create a new Pandas series that is a concatenation of the `Date` and `Time` series. In other words, let's concatenate the values stored in the `Date` and `Time` columns and store them into a new Pandas series. 

We will then convert these concatenated values to the `datetime` values using the `pd.to_datetime()` function from the Pandas module. The input to the `pd.to_datetime()` function is the Pandas series whose values are to be converted into `datetime` values.

**Syntax:** `pd.to_datetime(series)` 

where the `series` is the Pandas series whose values are to be converted into `datetime` values.

The default format of a `datetime` value is `YYYY-MM-DD` where `YYYY` denotes year, `MM` denotes month and `DD` denotes day.

But first we will have to replace the periods (or dots) with colons (`:`) in the values of the `Time` column because the time values are separated colons, i.e., `HH:MM:SS` where 

- `HH` denotes hours 

- `MM` denotes minutes

- `SS` denotes seconds

In [ ]:
time_series = pd.Series([item.replace(".",":") for item in df["Time"]], index = df.index)
time_series

0       18:00:00
1       19:00:00
2       20:00:00
3       21:00:00
4       22:00:00
          ...   
9352    10:00:00
9353    11:00:00
9354    12:00:00
9355    13:00:00
9356    14:00:00
Length: 9357, dtype: object

In [ ]:
date_series=pd.Series([item.split("/")[2]+"-"+item.split("/")[1]+"-"+item.split("/")[0] for item in df["Date"]] , index = df.index)
date_series

0       2004-03-10
1       2004-03-10
2       2004-03-10
3       2004-03-10
4       2004-03-10
           ...    
9352    2005-04-04
9353    2005-04-04
9354    2005-04-04
9355    2005-04-04
9356    2005-04-04
Length: 9357, dtype: object

In [ ]:
# S3.1: Concatenate the values stored in the 'Date' and 'Time' columns and store them into a new Pandas series.
dt_series = date_series + " "+time_series
dt_series = pd.to_datetime(dt_series)
dt_series

0      2004-03-10 18:00:00
1      2004-03-10 19:00:00
2      2004-03-10 20:00:00
3      2004-03-10 21:00:00
4      2004-03-10 22:00:00
               ...        
9352   2005-04-04 10:00:00
9353   2005-04-04 11:00:00
9354   2005-04-04 12:00:00
9355   2005-04-04 13:00:00
9356   2005-04-04 14:00:00
Length: 9357, dtype: datetime64[ns]

In the above code:

1. We iterate through each item of the `Time` column, convert it into a string, replace the period with the colon and then add the item to a Python list using the list comprehension method. We then convert the list into a Pandas series by applying the `pd.Series()` function. Inside the `pd.Series()` function, we also passed the `index` parameter whose value is `df.index`. This is to ensure that the indices of the new series are the same as the indices of the `df` DataFrame.

2. We concatenated the `Date` series with the new `Time` series using the `+` operator.

3. Finally, we converted the values of the newly concatenated series to the `datetime` values using the `pd.to_datetime()` function.

**Note:** The `pd.to_datetime()` returns a series containing `datetime64[ns]` values. They essentially are `datetime` values which Python reads as a combination of date and time.

---

#### Task 4: Parsing `datetime` Values - Other Functions

Let's learn how to extract date, day, month, year, time, hours, minutes, and seconds from a `datetime` value.

**The `date()` Function**

In general, a `datetime` value is a `Timestamp` object. It holds both date and time: 

In [ ]:
# S4.1: Get the first 'datetime' value from the 'df_series' and store it in the 'timestamp_0' variable. Print its value.
timestamp_0 = dt_series[0]
print(f"{timestamp_0} type = {type(timestamp_0)}")

2004-03-10 18:00:00 type = <class 'pandas._libs.tslibs.timestamps.Timestamp'>


To get a date (i.e., `datetime.date` object) from a timestamp, apply the `date()` function.

In [ ]:
# S4.2: Get date from the timestamp stored in the 'timestamp_0' variable. Store the output in the 'date_0' variable. 
# Print the value and its data-type.
date_0=timestamp_0.date()
print(date_0)

2004-03-10


**The `day, month` & `year` Attributes**

To get day, month and year from a `datetime.date` value apply the `day, month` and `year` attributes.

The output will be day, month and year as integer values.

In [ ]:
# S4.3: Get the day, month and year attributes from the date stored in the 'date_0' variable. Print the attributes and their data-types.
print(date_0.day)
print(date_0.month)
print(date_0.year)

10
3
2004


**The `strftime()` Function**

To convert a `datetime.date` value into a string value, use the `strftime()` function. You can also format the date in different formats using the `strftime()` function.

Let's format the date stored in the `date_0` variable in the `10-Mar-2004` format. To do this, you will have to pass the following input to the `strftime()` function.

`'%d-%b-%Y'` 

It is a combination of three format codes. They are:

1. `%d` format code denotes day of the month as a zero-padded decimal number. E.g., 01, 02, 03 etc.

2. `%b` format code denotes abbreviated month name of month in the `datetime.date` value. E.g., Jan Feb, March etc.

3. `%Y` format code denotes year with century as a decimal number. E.g, 2019, 2020, 2021 etc.

The hyphen `-` used here just to separate the day, month and year values. You can separate them with different characters.

To get a list of all the `strftime()` format codes, click on the link provided below and then go to the **`strftime()` and `strptime() Format Codes`** section.

https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes



In [ ]:
# S4.4: Format the date stored in the 'date_0' variable in the '10-Mar-2004' format. Print newly formatted date and also its data-type.
print(date_0.strftime("%d-%b-%Y"))

10-Mar-2004


In [ ]:
# S4.5: Format the date stored in the 'date_0' variable in the 'March 10, 2004' format. Print newly formatted date and also its data-type.
print(date_0.strftime("%B %d, %Y"))


March 10, 2004


**The `time()` Function**

To get time from a timestamp object, apply the `time()` function. It returns a `datetime.time` object.

In [ ]:
# S4.6: Get the time from the timestamp stored in the 'timestamp_0' variable. Store the output in the 'time_0' variable.
time_0=timestamp_0.time()
print(time_0)


18:00:00


**The `hour, minute` & `second` Attributes**

To get hour, minute and second values from a `datetime.time` object, apply the `hour, minute` and `second` attributes.

The output will be hour, minute and second as integer values.

In [ ]:
# S4.7: Get the hour, minute and second attributes from the time stored in the 'time_0' variable. Print the attributes and their data-types.
print(time_0.hour)
print(time_0.minute)
print(time_0.second)

18
0
0


Let's pause here. In the next class, we will learn more about `datetime` values.

---